### 1. Imports

In [129]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [130]:
############### Caminho das Bases ###############
goals_scorers = 'https://raw.githubusercontent.com/ErikHenning927/project_modulo3/main/goalscorers.csv'
results = 'https://raw.githubusercontent.com/ErikHenning927/project_modulo3/main/results.csv'
shootouts = 'https://raw.githubusercontent.com/ErikHenning927/project_modulo3/main/shootouts.csv'

############### Lendo as Bases ###############

df_goals = pd.read_csv(goals_scorers)
df_results = pd.read_csv(results)
df_shootouts = pd.read_csv(shootouts)

In [131]:
print("Colunas do df_goals:")
print(df_goals.columns)

print("\nColunas do df_results:")
print(df_results.columns)

print("\nColunas do df_shootouts:")
print(df_shootouts.columns)

Colunas do df_goals:
Index(['date', 'home_team', 'away_team', 'team', 'scorer', 'minute',
       'own_goal', 'penalty'],
      dtype='object')

Colunas do df_results:
Index(['date', 'home_team', 'away_team', 'home_score', 'away_score',
       'tournament', 'city', 'country', 'neutral'],
      dtype='object')

Colunas do df_shootouts:
Index(['date', 'home_team', 'away_team', 'winner', 'first_shooter'], dtype='object')


### 2. Renomeando as colunas

In [149]:
df_goals.rename(columns={
    'date': 'Data',
    'home_team': 'time_casa',
    'away_team': 'time_visitante',
    'team': 'Time_marcador',
    'scorer': 'artilheiro',
    'minute': 'minuto',
    'own_goal': 'gol_contra',
    'penalty': 'penalti'
}, inplace=True)

df_results.rename(columns={
    'date': 'Data',
    'home_team': 'Time_casa',
    'away_team': 'Time_visitante',
    'home_score': 'Gol_emcasa',
    'away_score': 'Gol_fora',
    'tournament': 'Torneio',
    'city': 'Cidade',
    'country': 'País',
    'neutral': 'Neutro'
 
}, inplace=True)
df_shootouts.rename(columns={
    'date': 'Data',
    'home_team': 'Time_casa',
    'away_team': 'Time_visitante',
    'winner': 'Vencedor',
    'first_shooter': 'Primeiro_Time_a_chutar_ao_Gol'
   
}, inplace=True)


             Data    time_casa time_visitante Time_marcador  \
0      02-07-1916        Chile        Uruguay       Uruguay   
1      02-07-1916        Chile        Uruguay       Uruguay   
2      02-07-1916        Chile        Uruguay       Uruguay   
3      02-07-1916        Chile        Uruguay       Uruguay   
4      06-07-1916    Argentina          Chile     Argentina   
...           ...          ...            ...           ...   
44105  10-02-2024        Qatar         Jordan         Qatar   
44106  10-02-2024        Qatar         Jordan         Qatar   
44107  11-02-2024  Ivory Coast        Nigeria       Nigeria   
44108  11-02-2024  Ivory Coast        Nigeria   Ivory Coast   
44109  11-02-2024  Ivory Coast        Nigeria   Ivory Coast   

                 artilheiro  minuto gol_contra penalti  
0           José Piendibene    44.0      False   False  
1          Isabelino Gradín    55.0      False   False  
2          Isabelino Gradín    70.0      False   False  
3           Jos

In [133]:
## Função para verificar o formato de data de uma coluna
def verificar_formato_data(df, coluna):
    primeiros_valores = df[coluna].head().tolist()
    # Tenta inferir o formato da data com base nos primeiros valores
    formatos = ["%Y-%m-%d", "%d-%m-%Y", "%m-%d-%Y", "%Y/%m/%d", "%d/%m/%Y", "%m/%d/%Y"]
    for formato in formatos:
        try:
            pd.to_datetime(primeiros_valores, format=formato)
            return formato
        except ValueError:
            continue
    return "Formato não reconhecido"

# Verificar o formato de data para cada coluna
print("Formato de data para df_goals['data']: ", verificar_formato_data(df_goals, 'Data'))
print("Formato de data para df_results['Data']: ", verificar_formato_data(df_results, 'Data'))
print("Formato de data para df_shootouts['Data']: ", verificar_formato_data(df_shootouts, 'Data'))


Formato de data para df_goals['data']:  %Y-%m-%d
Formato de data para df_results['Data']:  %Y-%m-%d
Formato de data para df_shootouts['Data']:  %Y-%m-%d


### 2. Tratamento de Dados (Conversão para tipos adequados)

In [ ]:
# Tratamento para a coluna 'data' em df_goals
df_goals['Data'] = pd.to_datetime(df_goals['Data'], format='%Y-%m-%d').dt.strftime('%d-%m-%Y')

# Tratamento para a coluna 'Data' em df_results
df_results['Data'] = pd.to_datetime(df_results['Data'], format='%Y-%m-%d').dt.strftime('%d-%m-%Y')

# Tratamento para a coluna 'Data' em df_shootouts
df_shootouts['Data'] = pd.to_datetime(df_shootouts['Data'], format='%Y-%m-%d').dt.strftime('%d-%m-%Y')

# Conversão da coluna 'minuto' para float em df_goals
df_goals['minuto'] = df_goals['minuto'].astype(float)

# Conversão das colunas 'Gol_emcasa' e 'Gol_fora' para int em df_results
df_results['Gol_emcasa'] = df_results['Gol_emcasa'].astype(int)
df_results['Gol_fora'] = df_results['Gol_fora'].astype(int)

# Imprimir as colunas tratadas
print(df_goals['Data'])
print(df_results['Data'])


### 3. Tratamento de Dados (removendo duplicados e NaN)

In [151]:
df_goals.dropna(inplace=True)
df_results.dropna(inplace=True)
df_shootouts.dropna(inplace=True)


df_goals.drop_duplicates(inplace=True)
df_results.drop_duplicates(inplace=True)
df_shootouts.drop_duplicates(inplace=True)


In [139]:
df_results[df_results['Gol_emcasa'] == df_results['Gol_fora']].shape[0]

10555

In [144]:
##Dataframe tratado
df_goals.head(3)

Data Time_casa Time_visitante  Gol_emcasa  Gol_fora   Torneio  \
0  30-11-1872  Scotland        England           0         0  Friendly   
1  08-03-1873   England       Scotland           4         2  Friendly   
2  07-03-1874  Scotland        England           2         1  Friendly   

    Cidade      País  Neutro  
0  Glasgow  Scotland   False  
1   London   England   False  
2  Glasgow  Scotland   False

In [141]:
##Dataframe tratado
df_results.head(3)

Data Time_casa Time_visitante  Gol_emcasa  Gol_fora   Torneio  \
0  30-11-1872  Scotland        England           0         0  Friendly   
1  08-03-1873   England       Scotland           4         2  Friendly   
2  07-03-1874  Scotland        England           2         1  Friendly   

    Cidade      País  Neutro  
0  Glasgow  Scotland   False  
1   London   England   False  
2  Glasgow  Scotland   False

In [142]:
##Dataframe tratado
df_shootouts.head(3)

Data       Time_casa  Time_visitante        Vencedor  \
26  20-06-1976  Czechoslovakia         Germany  Czechoslovakia   
36  22-05-1979       Argentina     Netherlands       Argentina   
40  21-06-1980           Italy  Czechoslovakia  Czechoslovakia   

   Primeiro_Time_a_chutar_ao_Gol  
26                Czechoslovakia  
36                     Argentina  
40                         Italy

## Levantamento de Inferências com base nos Dados:

**Questões:**

1. Qual é a média de gols marcados por jogo?

2. Quem são os 5 maiores artilheiros e quantos gols cada um marcou?

3. Qual país sediou o maior número de partidas?

4. Qual o desempenho médio de gols do Brasil em casa e fora nos últimos 10 anos.

5. Qual a correlação entre a equipe que inicia marcando e a probabilidade de vencer a partida?


>### Qual é a média de gols marcados por jogo?

In [145]:
df_results

Data     Time_casa Time_visitante  Gol_emcasa  Gol_fora  \
0      30-11-1872      Scotland        England           0         0   
1      08-03-1873       England       Scotland           4         2   
2      07-03-1874      Scotland        England           2         1   
3      06-03-1875       England       Scotland           2         2   
4      04-03-1876      Scotland        England           3         0   
...           ...           ...            ...         ...       ...   
46284  07-02-2024   Ivory Coast       DR Congo           1         0   
46285  07-02-2024       Nigeria   South Africa           1         1   
46286  10-02-2024         Qatar         Jordan           3         1   
46287  10-02-2024  South Africa       DR Congo           0         0   
46288  11-02-2024   Ivory Coast        Nigeria           2         1   

                      Torneio   Cidade         País  Neutro  
0                    Friendly  Glasgow     Scotland   False  
1                    Friendly   London      England   False  
2                    Friendly  Glasgow     Scotland   False  
3                    Friendly   London      England   False  
4                    Friendly  Glasgow     Scotland   False  
...                       ...      ...          ...     ...  
46284  African Cup of Nations   Ebimpé  Ivory Coast   False  
46285  African Cup of Nations   Bouaké  Ivory Coast    True  
46286           AFC Asian Cup   Lusail        Qatar   False  
46287  African Cup of Nations  Abidjan  Ivory Coast    True  
46288  African Cup of Nations   Ebimpé  Ivory Coast   False  

[46289 rows x 9 columns]

In [146]:
(df_results['Gol_emcasa'].sum() + df_results['Gol_fora'].sum()) / df_results.shape[0]

2.9370908855235585

> Ou seja, sabendo que temos uma quantidade de 46.289 resultados registrados, e somadas os gols fora e em casa temos uma média de aproximadamente ± 3 gols por jogo (2.93).


>### Quais são os 5 maiores marcadores e qual seu país de vínculo?

In [152]:
top_5_artilheiros = df_goals.groupby(['artilheiro', 'Time_marcador']).size().nlargest(5)
top_5_artilheiros

artilheiro          Time_marcador
Cristiano Ronaldo   Portugal         108
Robert Lewandowski  Poland            61
Romelu Lukaku       Belgium           60
Harry Kane          England           55
Lionel Messi        Argentina         54
dtype: int64

> O maior marcador é Cristiano Ronaldo, o português acumula 108 gols.


>### Qual País sediou mais partidas?

In [ ]:
print(f"O país com o maior número de partidas é: {df_results['País'].value_counts().idxmax()}, com um total de {df_results['País'].value_counts().max()} partidas.")


> O país com o maior número de partidas é: United States, com um total de 1339 partidas.


>### Qual o desempenho do brasil em casa e fora em 10 anos?

In [190]:
# Definir a data atual
from datetime import datetime, timedelta
data_atual = datetime.now()
intervalo = data_atual - timedelta(days=365*10) #10 anos

jogoscasa_brasil = df_results[df_results['Time_casa'] == 'Brazil']
# Filtrar com base no intervalo.
jogosbrasil = jogoscasa_brasil[pd.to_datetime(jogoscasa_brasil['Data'], format='%d-%m-%Y').dt.year >= intervalo.year]

# Calcular a média de gols marcados | casa
avg_goals_casa = jogosbrasil.groupby('Data')['Gol_emcasa'].mean()



In [191]:
avg_goals_casa

Data
01-02-2022    4.0
02-07-2018    2.0
02-07-2019    2.0
02-07-2021    1.0
03-06-2014    4.0
             ... 
28-03-2017    3.0
28-06-2014    1.0
28-11-2022    1.0
29-03-2015    1.0
31-08-2017    2.0
Name: Gol_emcasa, Length: 77, dtype: float64

In [197]:
# Definir a data atual
jogosfora_brasil = df_results[df_results['Time_visitante'] == 'Brazil']
# Filtrar com base no intervalo.
jogosbrasil = jogosfora_brasil[pd.to_datetime(jogosfora_brasil['Data'], format='%d-%m-%Y').dt.year >= intervalo.year]

# Calcular a média de gols marcados | casa
avg_goals_fora = jogosbrasil.groupby('Data')['Gol_fora'].mean()



In [198]:
avg_goals_fora

Data
01-09-2016    3.0
02-06-2022    5.0
02-09-2021    1.0
02-12-2022    0.0
03-06-2018    2.0
05-03-2014    5.0
05-09-2015    1.0
05-09-2017    1.0
05-10-2017    0.0
06-06-2022    1.0
07-09-2018    2.0
07-10-2021    3.0
08-06-2021    2.0
08-09-2015    4.0
08-10-2015    0.0
09-06-2017    0.0
09-12-2022    1.0
10-06-2018    3.0
10-10-2021    0.0
10-11-2017    3.0
11-10-2014    2.0
11-10-2016    2.0
12-09-2023    1.0
12-10-2018    2.0
12-11-2014    4.0
13-06-2017    4.0
13-10-2020    4.0
13-11-2015    1.0
14-10-2014    4.0
14-11-2017    0.0
15-11-2016    2.0
16-10-2018    1.0
16-11-2021    0.0
16-11-2023    1.0
17-10-2023    0.0
17-11-2020    2.0
18-11-2014    2.0
23-03-2017    4.0
23-03-2018    3.0
25-03-2023    1.0
26-03-2015    3.0
26-03-2019    3.0
27-01-2022    1.0
27-03-2018    1.0
27-06-2018    2.0
29-03-2016    2.0
29-03-2022    4.0
Name: Gol_fora, dtype: float64

>### Qual a correlação entre a equipe que inicia marcando e a probabilidade de vencer a partida?

In [209]:
first_shooter_counts = df_goals['Time_marcador'].value_counts()

vitoria_por_primeiro_marcador = df_shootouts.groupby('Primeiro_Time_a_chutar_ao_Gol')['Vencedor'].apply(lambda x: (x == x.name).mean())
# Calcular a correlação em porcentagem entre a contagem de primeiros chutes e a probabilidade de vitória
correlacao = first_shooter_counts.corr(vitoria_por_primeiro_marcador) * 100
print(f"Correlação entre o primeiro time a marcar e a probabilidade de vencer: {correlacao:.2f}%")

Correlação entre o primeiro time a marcar e a probabilidade de vencer: 6.31%
